# A1 # 

In [4]:
# load the list
T = [3007, 3067, 3244, 3375,3574, 3828, 4146, 4336, 4696, 5032,5234,5609,6094,6726,7226,7801,8592,9453,10565,11674,12575,13976,14434,15544,17121,18237,19071,20039,21417,22857,23889,24342,25419,26387,27695,28691,29968,31459,32854,34515,36330,37134,37998,39490,41725,44123,46302,48050,48570,47195,48651,50066,51784,53291,55124,56763,57867,59915,62805,65095,63028,69288]

In [5]:
# Question A1.1: Create a new list for GDP_increase
GDP_increase = [T[i+1]-T[i] for i in range (len(T)-1)]
print(GDP_increase)

[60, 177, 131, 199, 254, 318, 190, 360, 336, 202, 375, 485, 632, 500, 575, 791, 861, 1112, 1109, 901, 1401, 458, 1110, 1577, 1116, 834, 968, 1378, 1440, 1032, 453, 1077, 968, 1308, 996, 1277, 1491, 1395, 1661, 1815, 804, 864, 1492, 2235, 2398, 2179, 1748, 520, -1375, 1456, 1415, 1718, 1507, 1833, 1639, 1104, 2048, 2890, 2290, -2067, 6260]


## A1.1 ##

**GDP_increase: [60, 177, 131, 199, 254, 318, 190, 360, 336, 202, 375, 485, 632, 500, 575, 791, 861, 1112, 1109, 901, 1401, 458, 1110, 1577, 1116, 834, 968, 1378, 1440, 1032, 453, 1077, 968, 1308, 996, 1277, 1491, 1395, 1661, 1815, 804, 864, 1492, 2235, 2398, 2179, 1748, 520, -1375, 1456, 1415, 1718, 1507, 1833, 1639, 1104, 2048, 2890, 2290, -2067, 6260]**

In [6]:
# Question A1.2: Use GDP_increase to find the maximum gap size, i.e, the maximum increase in GDP per capita
Maximum = max(GDP_increase)
Maximum

6260

## A1.2 ##
**Maximum = 6260.**

In [7]:
# Question A.1.3: Find the percentage of gaps that have size greater than $1000.
GDP_greater_than_1000 = [num for num in GDP_increase if num > 1000]
Percentage_GDP_greater_than_1000 = (len(GDP_greater_than_1000)/len(GDP_increase))*100
Percentage_GDP_greater_than_1000

52.459016393442624

## A1.3 ##
**Percentage_GDP_greater_than_1000 = 52.459016393442624%**

In [8]:
# Question A.1.4: List of years in which the GDP per capita increase was more than $2000 
# The list starts from 1960 to 2021. Because of len(GDP_increase) + 1 = len(T). We need to restort the years count.

years_with_high_increase = [1960 + i + 1 for i, increase in enumerate(GDP_increase) if increase > 2000]
years_with_high_increase_alt = [1960 + i for i, increase in enumerate(GDP_increase, start=1) if increase > 2000]

# i is the index, returned by  enumerate(GDP_increase). i starts from 0 for index
# have start =1 inside of enumerate call. 
print(years_with_high_increase)
print(years_with_high_increase_alt)

[2004, 2005, 2006, 2017, 2018, 2019, 2021]
[2004, 2005, 2006, 2017, 2018, 2019, 2021]


## A1.4 ##

**years_with_high_increase = [2004, 2005, 2006, 2017, 2018, 2019, 2021].**

In [9]:
# Question A.1.5: Difference between the maximum and minimum GDP per capita values for each of the 5 year-periods & find increase minimum
# Find the index of 1976 
index_1976 = 1976-1960 
#Iterates through T in 5-year chunks, starting from the index corresponding to 1976 with step value = 5. 
differences = [max(T[i:i+5]) - min(T[i:i+5]) for i in range(index_1976, len(T)-4, 5)]
min_diff = min(differences)

min_diff_period_start = 1976 + 5 * differences.index(min_diff)
min_diff_period_end = min_diff_period_start + 4

print(f"Differences for each 5-year period starting from 1976: {differences}")
print(f"The 5-year period with the smallest difference was {min_diff_period_start}-{min_diff_period_end}, with a difference of {min_diff}.")



Differences for each 5-year period starting from 1976: [3983, 4261, 4818, 4349, 6362, 6989, 2349, 6697, 7228]
The 5-year period with the smallest difference was 2006-2010, with a difference of 2349.


## A1.5 ##
### A1.5.1 ###

**Differences for each 5-year period starting from 1976: [3983, 4261, 4818, 4349, 6362, 6989, 2349, 6697, 7228].**
### A1.5.2 ###

**The 5-year period with the smallest difference was 2006-2010, with a difference of 2349.**

# A2 #

In [10]:
majors_minors = majors_minors = [['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Humanities / Communications',  'Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Humanities / Communications',  'Social Sciences / Education',  'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Physical Sciences / Natural Sciences / Engineering'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Cognitive Science'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Music'], ['Social Sciences / Education'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Data Science'], ['Social Sciences / Education'], ['Math / Statistics / Computer Science', 'jazz'], ['Humanities / Communications', 'Social Sciences / Education'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Humanities / Communications',  'Social Sciences / Education',  'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Math / Statistics / Computer Science'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Econ'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', ''], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Humanities / Communications',  'Social Sciences / Education',  'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Humanities / Communications', 'Social Sciences / Education'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Social Sciences / Education'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Humanities / Communications', 'Social Sciences / Education'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education'], ['Humanities / Communications'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education',  'Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Humanities / Communications', 'Social Sciences / Education'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Humanities / Communications',  'Social Sciences / Education',  'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Humanities / Communications'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Social Sciences / Education',  'Physical Sciences / Natural Sciences / Engineering'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science', 'Music'], ['Physical Sciences / Natural Sciences / Engineering'], ['Humanities / Communications'], ['Physical Sciences / Natural Sciences / Engineering'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education', 'Math / Statistics / Computer Science'], ['Humanities / Communications', 'Math / Statistics / Computer Science'], ['Physical Sciences / Natural Sciences / Engineering'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science'], ['Social Sciences / Education'], ['Physical Sciences / Natural Sciences / Engineering'], ['Physical Sciences / Natural Sciences / Engineering',  'Math / Statistics / Computer Science'], ['Math / Statistics / Computer Science']]

In [11]:
# A2.1: Students have major / minor in any three of the above mentioned four areas
three_majors = [major_count for major_count in majors_minors if len(major_count) == 3]
len(three_majors)

6

## A2.1 ## 
**6 students have major/minor in any three of the four mentioned areas.**


In [12]:
# A2.2: students have Math / Statistics / Computer Science as an area of their major / minor
math_major = [math_count for math_count in majors_minors if "Math / Statistics / Computer Science" in math_count]
len(math_major)

105

## A2.2 ## 
**105 students have major/minor in Math / Statistics / Computer Science areas.**

In [13]:
# A2.3: Students only have Math / Statistics / Computer Science as an area of their major / minor
only_math_major = [student for student in majors_minors if len(student) == 1 and "Math / Statistics / Computer Science" in student]
len(only_math_major)

44

## A2.3 ## 
**44 students have major/minor only in Math / Statistics / Computer Science areas.**

In [14]:
# A2.4: Using list comprehension to find students with only "Math / Statistics / Computer Science" and "Social Sciences / Education"
desired_majors = [
    "Math / Statistics / Computer Science",
    "Social Sciences / Education"]

math_and_social_sci_dual = [
    student for student in majors_minors 
    if all(major in student for major in desired_majors) and len(student) == 2]

math_and_social_sci = [
    student for student in majors_minors 
    if all(major in student for major in desired_majors)]

print(len(math_and_social_sci_dual))
print(len(math_and_social_sci))


29
34


## A2.4 ## 
**29** students double major/minor in Math / Statistics / Computer Science and "Social Sciences / Education area.**

**34** students have majors/minors in Math / Statistics / Computer Science and "Social Sciences / Education area in thier academic plan (since 5  students have declared more than two major/minor in their selections.**

# A3 #  

In [15]:
starbucks_drinks_nutrition={'Cool Lime Starbucks Refreshers™ Beverage': [{'Nutrition_type': 'Calories', 'value': 45}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 11}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Strawberry Acai Starbucks Refreshers™ Beverage': [{'Nutrition_type': 'Calories', 'value': 80}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 18}, {'Nutrition_type': 'Fiber', 'value': 1}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Very Berry Hibiscus Starbucks Refreshers™ Beverage': [{'Nutrition_type': 'Calories', 'value': 60}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 14}, {'Nutrition_type': 'Fiber', 'value': 1}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Evolution Fresh™ Organic Ginger Limeade': [{'Nutrition_type': 'Calories', 'value': 110}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 28}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 5}], 'Iced Coffee': [{'Nutrition_type': 'Calories', 'value': 5}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 5}], 'Iced Espresso Classics - Vanilla Latte': [{'Nutrition_type': 'Calories', 'value': 130}, {'Nutrition_type': 'Fat', 'value': 2.5}, {'Nutrition_type': 'Carb', 'value': 21}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 5}, {'Nutrition_type': 'Sodium', 'value': 65}], 'Iced Espresso Classics - Caffe Mocha': [{'Nutrition_type': 'Calories', 'value': 140}, {'Nutrition_type': 'Fat', 'value': 2.5}, {'Nutrition_type': 'Carb', 'value': 23}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 5}, {'Nutrition_type': 'Sodium', 'value': 90}], 'Iced Espresso Classics - Caramel Macchiato': [{'Nutrition_type': 'Calories', 'value': 130}, {'Nutrition_type': 'Fat', 'value': 2.5}, {'Nutrition_type': 'Carb', 'value': 21}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 5}, {'Nutrition_type': 'Sodium', 'value': 65}], 'Shaken Sweet Tea': [{'Nutrition_type': 'Calories', 'value': 80}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 19}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Berry Blossom White': [{'Nutrition_type': 'Calories', 'value': 60}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 15}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Black Mango': [{'Nutrition_type': 'Calories', 'value': 150}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 38}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 15}], 'Tazo® Bottled Black with Lemon': [{'Nutrition_type': 'Calories', 'value': 140}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 35}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Brambleberry': [{'Nutrition_type': 'Calories', 'value': 140}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 35}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 15}], 'Tazo® Bottled Giant Peach': [{'Nutrition_type': 'Calories', 'value': 150}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 37}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 15}], 'Tazo® Bottled Iced Passion': [{'Nutrition_type': 'Calories', 'value': 70}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 17}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Lemon Ginger': [{'Nutrition_type': 'Calories', 'value': 120}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 31}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Organic Black Lemonade': [{'Nutrition_type': 'Calories', 'value': 140}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 35}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Organic Iced Black Tea': [{'Nutrition_type': 'Calories', 'value': 60}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 15}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Organic Iced Green Tea': [{'Nutrition_type': 'Calories', 'value': 120}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 31}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Plum Pomegranate': [{'Nutrition_type': 'Calories', 'value': 140}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 35}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Tazo® Bottled Tazoberry': [{'Nutrition_type': 'Calories', 'value': 150}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 38}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 15}], 'Tazo® Bottled White Cranberry': [{'Nutrition_type': 'Calories', 'value': 140}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 35}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Teavana® Shaken Iced Black Tea': [{'Nutrition_type': 'Calories', 'value': 30}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 8}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 5}], 'Teavana® Shaken Iced Black Tea Lemonade': [{'Nutrition_type': 'Calories', 'value': 70}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 17}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Teavana® Shaken Iced Green Tea': [{'Nutrition_type': 'Calories', 'value': 30}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 8}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 5}], 'Teavana® Shaken Iced Green Tea Lemonade': [{'Nutrition_type': 'Calories', 'value': 70}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 17}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 5}], 'Teavana® Shaken Iced Passion Tango™ Tea': [{'Nutrition_type': 'Calories', 'value': 30}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 8}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 5}], 'Teavana® Shaken Iced Passion Tango™ Tea Lemonade': [{'Nutrition_type': 'Calories', 'value': 90}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 24}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Teavana® Shaken Iced Peach Green Tea': [{'Nutrition_type': 'Calories', 'value': 60}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 15}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Starbucks Refreshers™ Raspberry Pomegranate': [{'Nutrition_type': 'Calories', 'value': 90}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 27}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Starbucks Refreshers™ Strawberry Lemonade': [{'Nutrition_type': 'Calories', 'value': 90}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 27}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Starbucks® Doubleshot Protein Dark Chocolate': [{'Nutrition_type': 'Calories', 'value': 210}, {'Nutrition_type': 'Fat', 'value': 2.5}, {'Nutrition_type': 'Carb', 'value': 33}, {'Nutrition_type': 'Fiber', 'value': 2}, {'Nutrition_type': 'Protein', 'value': 20}, {'Nutrition_type': 'Sodium', 'value': 115}], 'Starbucks® Doubleshot Protein Vanilla': [{'Nutrition_type': 'Calories', 'value': 200}, {'Nutrition_type': 'Fat', 'value': 2.5}, {'Nutrition_type': 'Carb', 'value': 34}, {'Nutrition_type': 'Fiber', 'value': 2}, {'Nutrition_type': 'Protein', 'value': 20}, {'Nutrition_type': 'Sodium', 'value': 120}], 'Starbucks® Iced Coffee Caramel': [{'Nutrition_type': 'Calories', 'value': 60}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 13}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Starbucks® Iced Coffee Light Sweetened': [{'Nutrition_type': 'Calories', 'value': 50}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 11}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Starbucks® Iced Coffee Unsweetened': [{'Nutrition_type': 'Calories', 'value': 10}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 2}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Blonde Roast': [{'Nutrition_type': 'Calories', 'value': 5}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Clover® Brewed Coffee': [{'Nutrition_type': 'Calories', 'value': 10}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Decaf Pike Place® Roast': [{'Nutrition_type': 'Calories', 'value': 5}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Featured Dark Roast': [{'Nutrition_type': 'Calories', 'value': 5}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Nariño 70 Cold Brew': [{'Nutrition_type': 'Calories', 'value': 5}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 15}], 'Nariño 70 Cold Brew with Milk': [{'Nutrition_type': 'Calories', 'value': 0}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Nitro Cold Brew': [{'Nutrition_type': 'Calories', 'value': 5}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 0}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Nitro Cold Brew with Sweet Cream': [{'Nutrition_type': 'Calories', 'value': 70}, {'Nutrition_type': 'Fat', 'value': 5.0}, {'Nutrition_type': 'Carb', 'value': 5}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 20}], 'Pike Place® Roast': [{'Nutrition_type': 'Calories', 'value': 5}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 0}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 10}], 'Vanilla Sweet Cream Cold Brew': [{'Nutrition_type': 'Calories', 'value': 110}, {'Nutrition_type': 'Fat', 'value': 6.0}, {'Nutrition_type': 'Carb', 'value': 14}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 1}, {'Nutrition_type': 'Sodium', 'value': 25}], 'Hot Chocolate': [{'Nutrition_type': 'Calories', 'value': 320}, {'Nutrition_type': 'Fat', 'value': 9.0}, {'Nutrition_type': 'Carb', 'value': 47}, {'Nutrition_type': 'Fiber', 'value': 4}, {'Nutrition_type': 'Protein', 'value': 14}, {'Nutrition_type': 'Sodium', 'value': 160}], 'Starbucks® Signature Hot Chocolate': [{'Nutrition_type': 'Calories', 'value': 430}, {'Nutrition_type': 'Fat', 'value': 26.0}, {'Nutrition_type': 'Carb', 'value': 45}, {'Nutrition_type': 'Fiber', 'value': 5}, {'Nutrition_type': 'Protein', 'value': 12}, {'Nutrition_type': 'Sodium', 'value': 115}], 'Caffè Latte': [{'Nutrition_type': 'Calories', 'value': 190}, {'Nutrition_type': 'Fat', 'value': 7.0}, {'Nutrition_type': 'Carb', 'value': 19}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 13}, {'Nutrition_type': 'Sodium', 'value': 170}], 'Caffè Mocha': [{'Nutrition_type': 'Calories', 'value': 290}, {'Nutrition_type': 'Fat', 'value': 8.0}, {'Nutrition_type': 'Carb', 'value': 42}, {'Nutrition_type': 'Fiber', 'value': 4}, {'Nutrition_type': 'Protein', 'value': 13}, {'Nutrition_type': 'Sodium', 'value': 140}], 'Cappuccino': [{'Nutrition_type': 'Calories', 'value': 120}, {'Nutrition_type': 'Fat', 'value': 4.0}, {'Nutrition_type': 'Carb', 'value': 12}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 8}, {'Nutrition_type': 'Sodium', 'value': 100}], 'Caramel Macchiato': [{'Nutrition_type': 'Calories', 'value': 250}, {'Nutrition_type': 'Fat', 'value': 7.0}, {'Nutrition_type': 'Carb', 'value': 35}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 10}, {'Nutrition_type': 'Sodium', 'value': 150}], 'Cinnamon Dolce Latte': [{'Nutrition_type': 'Calories', 'value': 260}, {'Nutrition_type': 'Fat', 'value': 6.0}, {'Nutrition_type': 'Carb', 'value': 40}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 11}, {'Nutrition_type': 'Sodium', 'value': 150}], 'Coconutmilk Mocha Macchiato': [{'Nutrition_type': 'Calories', 'value': 250}, {'Nutrition_type': 'Fat', 'value': 9.0}, {'Nutrition_type': 'Carb', 'value': 32}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 12}, {'Nutrition_type': 'Sodium', 'value': 180}], 'Flat White': [{'Nutrition_type': 'Calories', 'value': 180}, {'Nutrition_type': 'Fat', 'value': 7.0}, {'Nutrition_type': 'Carb', 'value': 18}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 12}, {'Nutrition_type': 'Sodium', 'value': 160}], 'Iced Caffè Latte': [{'Nutrition_type': 'Calories', 'value': 130}, {'Nutrition_type': 'Fat', 'value': 4.5}, {'Nutrition_type': 'Carb', 'value': 13}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 8}, {'Nutrition_type': 'Sodium', 'value': 115}], 'Iced Caffè Mocha': [{'Nutrition_type': 'Calories', 'value': 230}, {'Nutrition_type': 'Fat', 'value': 6.0}, {'Nutrition_type': 'Carb', 'value': 36}, {'Nutrition_type': 'Fiber', 'value': 4}, {'Nutrition_type': 'Protein', 'value': 9}, {'Nutrition_type': 'Sodium', 'value': 90}], 'Iced Caramel Macchiato': [{'Nutrition_type': 'Calories', 'value': 250}, {'Nutrition_type': 'Fat', 'value': 7.0}, {'Nutrition_type': 'Carb', 'value': 37}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 10}, {'Nutrition_type': 'Sodium', 'value': 150}], 'Iced Cinnamon Dolce Latte': [{'Nutrition_type': 'Calories', 'value': 200}, {'Nutrition_type': 'Fat', 'value': 4.0}, {'Nutrition_type': 'Carb', 'value': 34}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 7}, {'Nutrition_type': 'Sodium', 'value': 95}], 'Iced Coconutmilk Mocha Macchiato': [{'Nutrition_type': 'Calories', 'value': 260}, {'Nutrition_type': 'Fat', 'value': 9.0}, {'Nutrition_type': 'Carb', 'value': 34}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 11}, {'Nutrition_type': 'Sodium', 'value': 180}], 'Iced Vanilla Latte': [{'Nutrition_type': 'Calories', 'value': 190}, {'Nutrition_type': 'Fat', 'value': 4.0}, {'Nutrition_type': 'Carb', 'value': 30}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 7}, {'Nutrition_type': 'Sodium', 'value': 100}], 'Iced White Chocolate Mocha': [{'Nutrition_type': 'Calories', 'value': 300}, {'Nutrition_type': 'Fat', 'value': 8.0}, {'Nutrition_type': 'Carb', 'value': 47}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 10}, {'Nutrition_type': 'Sodium', 'value': 190}], 'Latte Macchiato': [{'Nutrition_type': 'Calories', 'value': 190}, {'Nutrition_type': 'Fat', 'value': 7.0}, {'Nutrition_type': 'Carb', 'value': 19}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 12}, {'Nutrition_type': 'Sodium', 'value': 160}], 'Starbucks Doubleshot® on Ice Beverage': [{'Nutrition_type': 'Calories', 'value': 45}, {'Nutrition_type': 'Fat', 'value': 1.0}, {'Nutrition_type': 'Carb', 'value': 5}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 3}, {'Nutrition_type': 'Sodium', 'value': 40}], 'Vanilla Latte': [{'Nutrition_type': 'Calories', 'value': 250}, {'Nutrition_type': 'Fat', 'value': 6.0}, {'Nutrition_type': 'Carb', 'value': 37}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 12}, {'Nutrition_type': 'Sodium', 'value': 150}], 'White Chocolate Mocha': [{'Nutrition_type': 'Calories', 'value': 360}, {'Nutrition_type': 'Fat', 'value': 11.0}, {'Nutrition_type': 'Carb', 'value': 53}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 14}, {'Nutrition_type': 'Sodium', 'value': 240}], 'Cinnamon Dolce Frappuccino® Blended Coffee': [{'Nutrition_type': 'Calories', 'value': 350}, {'Nutrition_type': 'Fat', 'value': 4.5}, {'Nutrition_type': 'Carb', 'value': 64}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 15}, {'Nutrition_type': 'Sodium', 'value': 0}], 'Coffee Light Frappuccino® Blended Coffee': [{'Nutrition_type': 'Calories', 'value': 110}, {'Nutrition_type': 'Fat', 'value': 0.0}, {'Nutrition_type': 'Carb', 'value': 24}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 3}, {'Nutrition_type': 'Sodium', 'value': 200}], 'Mocha Frappuccino® Blended Coffee': [{'Nutrition_type': 'Calories', 'value': 280}, {'Nutrition_type': 'Fat', 'value': 2.5}, {'Nutrition_type': 'Carb', 'value': 60}, {'Nutrition_type': 'Fiber', 'value': 2}, {'Nutrition_type': 'Protein', 'value': 4}, {'Nutrition_type': 'Sodium', 'value': 220}], 'Mocha Light Frappuccino® Blended Coffee': [{'Nutrition_type': 'Calories', 'value': 140}, {'Nutrition_type': 'Fat', 'value': 0.5}, {'Nutrition_type': 'Carb', 'value': 28}, {'Nutrition_type': 'Fiber', 'value': 1}, {'Nutrition_type': 'Protein', 'value': 4}, {'Nutrition_type': 'Sodium', 'value': 180}], 'Cinnamon Dolce Crème': [{'Nutrition_type': 'Calories', 'value': 200}, {'Nutrition_type': 'Fat', 'value': 6.0}, {'Nutrition_type': 'Carb', 'value': 28}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 10}, {'Nutrition_type': 'Sodium', 'value': 135}], 'Vanilla Crème': [{'Nutrition_type': 'Calories', 'value': 200}, {'Nutrition_type': 'Fat', 'value': 6.0}, {'Nutrition_type': 'Carb', 'value': 28}, {'Nutrition_type': 'Fiber', 'value': 0}, {'Nutrition_type': 'Protein', 'value': 10}, {'Nutrition_type': 'Sodium', 'value': 135}], 'Chocolate Smoothie': [{'Nutrition_type': 'Calories', 'value': 320}, {'Nutrition_type': 'Fat', 'value': 5.0}, {'Nutrition_type': 'Carb', 'value': 53}, {'Nutrition_type': 'Fiber', 'value': 8}, {'Nutrition_type': 'Protein', 'value': 20}, {'Nutrition_type': 'Sodium', 'value': 170}], 'Strawberry Smoothie': [{'Nutrition_type': 'Calories', 'value': 300}, {'Nutrition_type': 'Fat', 'value': 2.0}, {'Nutrition_type': 'Carb', 'value': 60}, {'Nutrition_type': 'Fiber', 'value': 7}, {'Nutrition_type': 'Protein', 'value': 16}, {'Nutrition_type': 'Sodium', 'value': 130}]}

In [16]:
# A3.1: The nested data structure of the object starbucks_drinks_nutrition
# Data structure: 
# starbucks_drinks_nutrition is a dictionary,  specific structure is listed below: 
starbucks_drinks_nutrition_structure = {
    'Drink Name 1': [
        {'Nutrition_type': 'Nutrition_type_name', 'value': 0.0},
        {'Nutrition_type': 'Nutrition_type_name', 'value': 0.0},
        # ... other nutrition info ...
    ],
    # ... other drinks ...
}



## A3.1 ##
starbucks_drinks_nutrition, is a dictionary where each key is a string representing the name of a Starbucks drink, and each value is a list of dictionaries. Each dictionary within the list contains two key-value pairs: 'Nutrition_type' and 'value', which represent the type of nutrition fact (e.g., Calories, Fat, Carb, etc.) and its corresponding value.

In [17]:
# A3.2: print the name and carb content of all drinks that have a carb content of more than 50 units
high_carb = []
for drink, nutrition in starbucks_drinks_nutrition.items(): 
    for type_nutrition in nutrition: 
        if type_nutrition['Nutrition_type'] == 'Carb' and type_nutrition['value'] > 50:
            high_carb.append((drink, type_nutrition['value'])) 


for drink, carb in high_carb:
    print(f"{drink}: {carb}g of carbohydrates")

White Chocolate Mocha: 53g of carbohydrates
Cinnamon Dolce Frappuccino® Blended Coffee: 64g of carbohydrates
Mocha Frappuccino® Blended Coffee: 60g of carbohydrates
Chocolate Smoothie: 53g of carbohydrates
Strawberry Smoothie: 60g of carbohydrates


## A3.2 ##

**White Chocolate Mocha**: 53g of carbohydrates

**Cinnamon Dolce Frappuccino® Blended Coffee**: 64g of carbohydrates

**Mocha Frappuccino® Blended Coffee**: 60g of carbohydrates

**Chocolate Smoothie**: 53g of carbohydrates

**Strawberry Smoothie**: 60g of carbohydrates

# A4 #

In [18]:
# A4.1: load data 
import pandas as pd
# load  external data
data = pd.read_json("https://raw.githubusercontent.com/cwkenwaysun/TEDmap/master/data/TED_Talks.json")
data.head()

id       speaker                              headline  \
0   7   David Pogue                      Simplicity sells   
1   6  Craig Venter              Sampling the ocean's DNA   
2   4    Burt Rutan  The real future of space exploration   
3   3  Ashraf Ghani         How to rebuild a broken state   
4   5  Chris Bangle              Great cars are great art   

                                  URL  \
0  http://www.ted.com/talks/view/id/7   
1  http://www.ted.com/talks/view/id/6   
2  http://www.ted.com/talks/view/id/4   
3  http://www.ted.com/talks/view/id/3   
4  http://www.ted.com/talks/view/id/5   

                                         description  \
0  New York Times columnist David Pogue takes aim...   
1  Genomics pioneer Craig Venter takes a break fr...   
2  In this passionate talk, legendary spacecraft ...   
3  Ashraf Ghani's passionate and powerful 10-minu...   
4  American designer Chris Bangle explains his ph...   

                                      transcript_URL  month_filmed  \
0  http://www.ted.com/talks/view/id/7/transcript?...             2   
1  http://www.ted.com/talks/view/id/6/transcript?...             7   
2  http://www.ted.com/talks/view/id/4/transcript?...             2   
3  http://www.ted.com/talks/view/id/3/transcript?...             7   
4  http://www.ted.com/talks/view/id/5/transcript?...             2   

   year_filmed           event duration date_published  \
0         2006         TED2006  0:21:26        6/27/06   
1         2005  TEDGlobal 2005  0:16:51     2004/05/07   
2         2006         TED2006  0:19:37       10/25/06   
3         2005  TEDGlobal 2005  0:18:45       10/18/06   
4         2002         TED2002  0:20:04     2004/05/07   

                                                tags  \
0  simplicity,computers,software,interface design...   
1  biotech,invention,oceans,genetics,DNA,biology,...   
2  aircraft,flight,industrial design,NASA,rocket ...   
3  corruption,poverty,economics,investment,milita...   
4  cars,industrial design,transportation,inventio...   

                                              newURL       date    views  \
0  https://www.ted.com/talks/david_pogue_says_sim... 2006-06-27  1646773   
1  https://www.ted.com/talks/craig_venter_on_dna_... 2004-05-07   562625   
2  https://www.ted.com/talks/burt_rutan_sees_the_... 2006-10-25  2046869   
3  https://www.ted.com/talks/ashraf_ghani_on_rebu... 2006-10-18   814554   
4  https://www.ted.com/talks/chris_bangle_says_gr... 2004-05-07   870950   

                                               rates  
0  [{'id': 7, 'name': 'Funny', 'count': 968}, {'i...  
1  [{'id': 3, 'name': 'Courageous', 'count': 21},...  
2  [{'id': 3, 'name': 'Courageous', 'count': 169}...  
3  [{'id': 3, 'name': 'Courageous', 'count': 140}...  
4  [{'id': 1, 'name': 'Beautiful', 'count': 89}, ...

## A4.1 ##
**Please find the above**

In [19]:
# 4.2: Find the total number of talks
data_2006_2017 = [(data['year_filmed'] >= 2006) & (data['year_filmed'] <= 2017)]
len(data)

2474

## A4.2 ##
**There are 2474 talks in the dataset**

In [20]:
#4.3: find the most viewed talk
max_views = data['views'].idxmax()
pop_talk = data.loc[max_views]

In [21]:
print(f"A4.3: The talk with the highest number of views has the following details:\n"
      f"Headline: {pop_talk ['headline']}\n"
      f"Speaker: {pop_talk ['speaker']}\n"
      f"Year Filmed: {pop_talk ['year_filmed']}")

A4.3: The talk with the highest number of views has the following details:
Headline: Do schools kill creativity?
Speaker: Ken Robinson
Year Filmed: 2006


## A4.3 ##
**The talk with the highest number of views has the following details:**

Headline: Do schools kill creativity?

Speaker: Ken Robinson 

Year Filmed: 2006

In [170]:
# 4.4: Compare with the average
average_views = data['views'].mean()
high_ave = data[data['views'] > average_views]
low_ave = data[data['views'] < average_views]

equal_view_count = sum(data['views'] == average_views)
print(equal_view_count)

if len(low_ave) > len(high_ave):
    print("The majority of talks have less views than average.")
else:
    print("The majority of talks have more views than the average.")

0
The majority of talks have less views than average.


## A4.4 ##
**The majority of talks have less views than average.**
The average number of views is found using mean() function. Then, I created two DataFrames: one for talks with below-average views and one for talks with above-average views. The majority is determined by comparing the lengths of these DataFrames, and the results are printed out with justification.

In [171]:
# A.4.5: 25% of the talks have more views than the average number of views for a talk?
higher_than_ave = (len(high_ave) / len(data)) * 100
higher_than_ave

25.060630557801133

## A4.5 ##
**Yes, more than 25% of the talks have higher views than the average.**
I divide the length of the dataset by  the length of the high_ave DataFrame. The ratio 25.060630557801133% is greater than 25%. Therefore, at least 25% of the talks have more views than the average number of views for a talk

In [172]:
# 4.6: create a new column in the dataset that contains the votes obtained by the tedtalk in that category. Print the first 5 rows of the updated dataset.
for index, row in data.iterrows():
    for rate_dict in row['rates']:
        # Extracting category and votes and creating new columns
        category = rate_dict['name']
        votes = rate_dict['count']
        data.loc[index, category] = votes

print(data.head())

   id       speaker                              headline  \
0   7   David Pogue                      Simplicity sells   
1   6  Craig Venter              Sampling the ocean's DNA   
2   4    Burt Rutan  The real future of space exploration   
3   3  Ashraf Ghani         How to rebuild a broken state   
4   5  Chris Bangle              Great cars are great art   

                                  URL  \
0  http://www.ted.com/talks/view/id/7   
1  http://www.ted.com/talks/view/id/6   
2  http://www.ted.com/talks/view/id/4   
3  http://www.ted.com/talks/view/id/3   
4  http://www.ted.com/talks/view/id/5   

                                         description  \
0  New York Times columnist David Pogue takes aim...   
1  Genomics pioneer Craig Venter takes a break fr...   
2  In this passionate talk, legendary spacecraft ...   
3  Ashraf Ghani's passionate and powerful 10-minu...   
4  American designer Chris Bangle explains his ph...   

                                      transcript_

## A4.6 ##
**The new columns corresponding to categories in rates are created above.**

In [173]:
# A4.7: The most confusing talk" 
idx_max_confusing = data['Confusing'].idxmax()
most_confusing_talk = data.loc[idx_max_confusing, 'headline']
most_confusing_talk

'I believe we evolved from aquatic apes'

## A.4.7 ## 
**The talk with the highest number of 'Confusing' votes is:I believe we evolved from aquatic apes**

In [174]:
# 4.8： Highest fascinating count 
def get_fascinating_percentage(rates):
    total_votes = sum(rate['count'] for rate in rates)
    fascinating_votes = next((rate['count'] for rate in rates if rate['name'] == 'Fascinating'), 0)
    return (fascinating_votes / total_votes) * 100 if total_votes > 0 else 0

data['fascinating_percentage'] = data['rates'].apply(get_fascinating_percentage)

highest_fascinating_talk = data.loc[data['fascinating_percentage'].idxmax()]
print(highest_fascinating_talk[['speaker', 'headline', 'fascinating_percentage', 'date']])

speaker                                      Carolyn Porco
headline                  Could a Saturn moon harbor life?
fascinating_percentage                           43.887423
date                                   2009-05-21 00:00:00
Name: 443, dtype: object


## A4.8 ##
The highest percentage of votes in the Fascinating category.


**Headline: Could a Saturn moon harbor life?**

**Year of the talk: 2009**

# A5 # 

In [175]:
# 5.1: load data
data = pd.read_csv('/Users/jinwens/Downloads/univ.txt', delimiter='\t')
print(data.head())

                    Name       Location  Rank  Tuition and fees  \
0   Princeton University  Princeton, NJ   1.0           45320.0   
1     Harvard University  Cambridge, MA   2.0           47074.0   
2  University of Chicago    Chicago, IL   3.0           52491.0   
3        Yale University  New Haven, CT   3.0           49480.0   
4    Columbia University   New York, NY   5.0           55056.0   

   Undergrad Enrollment  
0                5402.0  
1                6699.0  
2                5844.0  
3                5532.0  
4                6102.0  


## A5.1 ## 
**Data is loaded above.**

In [176]:
# 5.2: Find summary statistics of the data
print(data.describe())

             Rank  Tuition and fees  Undergrad Enrollment
count  231.000000        231.000000            231.000000
mean   113.982684      33769.246753          14946.619048
std     65.995518      10756.733516          10569.664095
min      1.000000       5300.000000           1001.000000
25%     56.000000      25693.000000           6238.500000
50%    111.000000      31608.000000          12949.000000
75%    171.000000      42721.000000          22145.500000
max    220.000000      55056.000000          54513.000000


## A5.2 ## 
**Summary of the data is presented above.**

In [177]:
# 5.3: How many universities are there in the data set?
num_universities = data.shape[0]
print(f'There are {num_universities} universities in the dataset.')

There are 231 universities in the dataset.


## A5.3 ##
**There are 231 universities in the dataset.**

In [178]:
# A5.4: Estimate the maximum Tuition and fees among universities that are in the bottom 25% when ranked by total tuition and fees.
bottom_25_threshold = data['Tuition and fees'].quantile(0.25)
max_tuition_bottom_25 = data[data['Tuition and fees'] <= bottom_25_threshold]['Tuition and fees'].max()
print(f'The maximum tuition and fees among the bottom 25% of universities is {max_tuition_bottom_25}.')

The maximum tuition and fees among the bottom 25% of universities is 25673.0.


## A5.4 ##

**The maximum tuition and fees among the bottom 25% of universities is 25673.0.**

In [179]:
# A5.5: How many universities share the ranking of 220
num_universities_rank_220 = data[data['Rank'] == 220].shape[0]
print(f'{num_universities_rank_220} universities share the ranking of 220.')

12 universities share the ranking of 220.


## A5.5 ## 
**12 universities share the ranking of 220.**

## A5.6 ##
No, determining the mean "Tuition and fees" specifically for undergraduate students in the US from the summary statistics is not feasible. The summary statistics merely provide an average of "Tuition and fees" across all educational levels at universities.

In [180]:
# A5.7: Find the average Tuition and fees for an undergrad student in the US.
total_tuition_fees = (data['Tuition and fees'] * data['Undergrad Enrollment']).sum()
total_undergrad_enrollment = data['Undergrad Enrollment'].sum()

average_tuition_per_undergrad = total_tuition_fees / total_undergrad_enrollment

# Display the result
print(f'The average Tuition and fees per undergrad student in the US is ${average_tuition_per_undergrad}.')

The average Tuition and fees per undergrad student in the US is $30845.32903298868.


## A5.7 ##
**The mean Tuition and fees for an undergrad student in the US is $30845.32903298868.**

# A6 #

#### 1. CSV (Comma-Separated Values)
   - **Advantage:** 
     - **Interoperability:** CSV files are universally recognized and can be opened by a wide range of software, from spreadsheet applications like Microsoft Excel to database systems. This makes it a popular choice for data exchange between different platforms.
   
   - **Disadvantage:** 
     - **Flat Structure:** CSV files inherently support only two-dimensional, tabular data. They cannot handle hierarchical or multi-dimensional data as efficiently as JSON. This makes them less suitable for complex data structures.

#### 2. JSON (JavaScript Object Notation)
   - **Advantage:** 
     - **Flexibility:** JSON supports hierarchical data structures, allowing for nested arrays and objects. This makes it suitable for representing complex data structures, especially in applications that require data interchange between a server and a web application.
   
   - **Disadvantage:** 
     - **Size:** For large datasets, JSON can be more verbose than CSV, leading to larger file sizes. This can be a concern when storage or data transmission efficiency is crucial.

#### 3. TXT (Text File)
   - **Advantage:** 
     - **Unstructured Freedom:** TXT files are free-form, allowing users to store data in any format. This can be useful for custom data storage needs or for human-readable notes and documents.
   
   - **Disadvantage:** 
     - **Lack of Standardization:** Since TXT files don't have a standard structure, parsing them can be challenging. Unlike CSV and JSON, which have well-defined formats, reading data from TXT files might require custom logic, especially if the structure isn't consistent.

